packages/asu-core/__tests__/aws/services/cache.service.spec.ts

```diff
@@ -0,0 +1,149 @@
+import { CacheProvider } from '../../../src/aws/providers/cache.provider';
+import { CacheService } from '../../../src/aws/services/cache.service';
+import { Mock } from "../../../src/testing/mock-provider";
+import * as redis from 'redis';
+import { TimeToLive } from '../../../src/aws/enums/time-to-live';
+
+describe('CacheService', () => {
+    const setup = () => {
+        const cacheProvider = Mock(new CacheProvider);
+        const cacheClient = Mock(redis.createClient());
+        const service = new CacheService(cacheProvider);
+
+        cacheProvider.resolve.mockReturnValue(cacheClient);
+
+        return { service, cacheClient, cacheProvider };
+    };
+
+    describe('getValue', () => {
+        it('should return cached value for key', async () => {
+            // Arrange
+            const { service, cacheClient } = setup();
+            const expectedKey = 'key';
+            const expectedValue = 'value';
+
+            cacheClient.get.mockResolvedValue(expectedValue);
+
+            // Act
+            const result = await service.getValue(expectedKey, async () => {
+                return null;
+            });
+
+            // Assert
+            expect(result).toBe(expectedValue);
+        });
+
+        it('should call given callback when cache fails to connect', async () => {
+            // Arrange
+            const { service, cacheClient } = setup();
+            const expectedKey = 'key';
+            const expectedValue = 'value';
+
+            cacheClient.connect.mockImplementation(() => {
+                throw 'connection timed out';
+            });
+
+            // Act
+            const result = await service.getValue(expectedKey, async () => {
+                return expectedValue;
+            });
+
+            // Assert
+            expect(result).toBe(expectedValue);
```

Comment:
Instead of just checking the expected output, what do you think about wrapping the callback with `jest.fn()` and asserting it was called?:
```
            const mockCallback = jest.fn(async () => {
                return expectedValue;
            })
            
            // Act
            const result = await service.getValue(expectedKey, mockCallback);

            // Assert
            expect(mockCallback).toHaveBeenCalled();
            expect(result).toBe(expectedValue);
```

Reply:
I love that idea. Implemented.